backend/app/routers/metadata_datasets.py

```diff
@@ -0,0 +1,260 @@
+from fastapi import APIRouter, HTTPException, Depends
+from fastapi import Form
+from pymongo import MongoClient
+
+from app import keycloak_auth
+from app import dependencies
+from app.keycloak_auth import get_user, get_current_user, UserOut
+from app.config import settings
+from app.models.datasets import DatasetOut
+from app.models.extractors import ExtractorIn
+from app.models.metadata import (
+    MongoDBRef,
+    MetadataAgent,
+    MetadataDefinitionOut,
+    MetadataIn,
+    MetadataDB,
+    MetadataOut,
+    MetadataPatch,
+    validate_context,
+    patch_metadata
+)
+
+router = APIRouter()
+
+clowder_bucket = os.getenv("MINIO_BUCKET_NAME", "clowder")
+
+
+async def _build_metadata_db_obj(db: MongoClient, metadata_in: MetadataIn, dataset: DatasetOut, user: UserOut,
+                                 agent: MetadataAgent = None):
+    contents = await validate_context(db, metadata_in.contents,
+                                              metadata_in.definition, metadata_in.context_url, metadata_in.context)
+
+    if agent is None:
+        # Build MetadataAgent depending on whether extractor info is present
+        extractor_info = metadata_in.extractor_info
+        if extractor_info is not None:
+            extractor_in = ExtractorIn(**extractor_info.dict())
+            if (
+                    extractor := await db["extractors"].find_one(
+                        {"_id": extractor_in.id, "version": extractor_in.version}
+                    )
+            ) is not None:
+                agent = MetadataAgent(creator=user, extractor=extractor)
+            else:
+                raise HTTPException(status_code=404, detail=f"Extractor not found")
+        else:
+            agent = MetadataAgent(creator=user)
+
+    dataset_ref = MongoDBRef(collection="datasets", resource_id=dataset["_id"])
+
+    # Apply any typecast fixes from definition validation
+    metadata_in = metadata_in.dict()
+    metadata_in["contents"] = contents
+    return MetadataDB(
+        **metadata_in,
+        resource=dataset_ref,
+        agent=agent,
+    )
+
+
+@router.post("/{dataset_id}/metadata", response_model=MetadataOut)
+async def add_dataset_metadata(
+    metadata_in: MetadataIn,
+    dataset_id: str,
+    user=Depends(get_current_user),
+    db: MongoClient = Depends(dependencies.get_db),
+):
+    """Attach new metadata to a dataset. The body must include a contents field with the JSON metadata, and either a
+    context JSON-LD object, context_url, or definition (name of a metadata definition) to be valid.
+
+    Returns:
+        Metadata document that was added to database
+    """
+    if (
+        dataset := await db["datasets"].find_one({"_id": ObjectId(dataset_id)})
+    ) is not None:
+        # If dataset already has metadata using this definition, don't allow duplication
+        definition = metadata_in.definition
+        if definition is not None:
+            existing_q = {
+                "resource.resource_id": dataset.id,
```

Comment:
I got an error complaing that dict object doesn't have id field and I looked at the object it has "_id" instead: 
<img width="1082" alt="image" src="https://user-images.githubusercontent.com/13950475/171204577-42ba0eaa-6101-4c89-b101-82ebdc5a8c78.png">

I could fix it on my machine but then i looked at similar endpoints on `metadata_file.py` and saw you are using file.id there and it worked... No sure what's going on here? @max-zilla

Reply:
Oh actually i know... it's missing `dataset = DatasetOut(**dataset)`.